data_model/derive/src/lib.rs

```diff
@@ -0,0 +1,198 @@
+//! Crate with `Filter` and `EventFilter` derive macro
+#![allow(clippy::expect_used, clippy::panic, clippy::too_many_lines)]
+
+use proc_macro::TokenStream;
+use quote::{format_ident, quote};
+
+/// [`Filter`] is used for code generation of `...Filter` structs and `...EventFilter` enums, as well as
+/// implementing the `Filter` trait for both of them. This macro should only be attributed to `Event` enums.
+/// Due to name scoping, the macro currently only
+/// expands properly from within the `iroha_data_model` crate as it relies on a few of `crate::prelude`
+/// imports. This macro is also dependent on the naming conventions adopted so far, such as that
+/// `Event` enums always have tuple variants with either some sort of `Id` or another `Event` inside
+/// of them.
+#[proc_macro_derive(Filter)]
+pub fn filter_derive(input: TokenStream) -> TokenStream {
+    let ast = syn::parse(input).expect("Failed to parse input TokenStream.");
+    impl_from_filter(&ast)
+}
+
+fn impl_from_filter(ast: &syn::DeriveInput) -> TokenStream {
+    // Omitting attributes as they don't need to be inherited for filters
+    let syn::DeriveInput {
+        vis,
+        ident: event_ident,
+        generics,
+        data,
+        ..
+    } = ast;
+
+    let event_filter_ident = format_ident!("{}Filter", event_ident);
+
+    let filter_ident = format_ident!(
+        "{}Filter",
+        event_ident
+            .to_string()
+            .strip_suffix("Event")
+            .expect("Events should follow the naming format")
```

Comment:
OK, so we need to call it `VariantEvent` and it becomes `VariantFilter`, right? This is a good convention. Some examples would be welcome, in the docs.